provider/provider.go

```diff
@@ -70,6 +78,50 @@ func NewPinnedProvider(onlyRoots bool, pinning pin.Pinner, fetchConfig fetcher.F
 	}
 }
 
+// NewBufferedProvider returns a KeyChanFunc supplying keys from a given
+// KeyChanFunction, but buffering all the keys in memory first until they are
+// read.  This allows the underlying KeyChanFunc to finish listing pins as
+// soon as possible releasing any resources, locks, at the expense of memory
+// usage.
+func NewBufferedProvider(pinsF KeyChanFunc) KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		pins, err := pinsF(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		// Use very efficient queue implementation for storing the cids.
+		queue := new(deque.Deque[cid.Cid])
+		queue.SetBaseCap(1024)
+
+		totalCids := 0
+		for c := range pins {
+			queue.PushBack(c)
+			totalCids++
+		}
+
+		outCh := make(chan cid.Cid) // channel for the consumer.
+
+		go func() {
+			defer close(outCh)
+			for {
+				if totalCids == 0 {
+					return
+				}
+				c := queue.PopFront()
+				select {
+				case <-ctx.Done():
+					return
+				case outCh <- c:
+					totalCids--
+				}
+			}
+		}()
+
+		return outCh, nil
+	}
```

Comment:
You may consider using [chanqueue](https://github.com/gammazero/chanqueue?tab=readme-ov-file#chanqueue) for this purpose, as it basically does what the code above does. A chanqueue can supply its own in and out channels or use channels provided to it for input and output.